src/main/java/org/pentaho/di/trans/steps/zendesk/ZendeskInputIncremental.java

```diff
@@ -87,44 +90,68 @@ public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws
+
     switch ( meta.getDownloadType() ) {
       case TICKETS:
-        for ( Ticket ticket : data.conn.getTicketsIncrementally( startDate ) ) {
-          putRow( data.rowMeta, processTicket( ticket ) );
-          incrementLinesOutput();
-          if ( isStopped() ) {
-            break;
-          }
-        }
+        serviceIterator = data.conn.getTicketsIncrementally( startDate ).iterator();
         break;
       case USERS:
-        for ( User user : data.conn.getUsersIncrementally( startDate ) ) {
-          putRow( data.rowMeta, processUser( user ) );
-          incrementLinesOutput();
-          if ( isStopped() ) {
-            break;
-          }
-        }
+        serviceIterator = data.conn.getUsersIncrementally( startDate ).iterator();
         break;
       case ORGANIZATIONS:
-        for ( Organization org : data.conn.getOrganizationsIncrementally( startDate ) ) {
-          putRow( data.rowMeta, processOrganization( org ) );
-          incrementLinesOutput();
-          if ( isStopped() ) {
-            break;
-          }
-        }
+        serviceIterator = data.conn.getOrganizationsIncrementally( startDate ).iterator();
         break;
       case HELPCENTER_ARTICLES:
-        for ( Article article : data.conn.getArticlesIncrementally( startDate ) ) {
-          putRow( data.rowMeta, processArticle( article ) );
-          incrementLinesOutput();
-          if ( isStopped() ) {
-            break;
-          }
-        }
+        serviceIterator = data.conn.getArticlesIncrementally( startDate ).iterator();
         break;
-    }
+      default:
+        serviceIterator = null; //get rid of compile error
+        break;
+     }
+     allrecords:
+     while(true) {
+       try {
+         while (serviceIterator.hasNext() ) {
+           switch ( meta.getDownloadType() ) {
+             case TICKETS:
+               Ticket ticket=(Ticket)(serviceIterator.next());
+               putRow( data.rowMeta, processTicket( ticket ) );
+               break;
+             case USERS:
+               User user=(User)(serviceIterator.next());
+               putRow( data.rowMeta, processUser( user ) );
+               break;
+             case ORGANIZATIONS:
+               Organization org=(Organization)(serviceIterator.next());
+               putRow( data.rowMeta, processOrganization( org ) );
+               break;
+             case HELPCENTER_ARTICLES:
+               Article article=(Article)(serviceIterator.next());
+               putRow( data.rowMeta, processArticle( article ) );
+               break;
+           }
+           incrementLinesOutput();
+           if ( isStopped() ) {
+             break allrecords;
+           }
+         }
+         break allrecords; // We need to exit from the upper loop now we have all records
+       } catch ( ZendeskResponseRateLimitException zre ) {
+         Long retryAfter = zre.getRetryAfter();
+         logBasic ( "Hit rate limiting. Sleeping " + retryAfter + "s" );
```

Comment:
Please use the BaseMessages API here, so that multiple languages can easily be supported.

Reply:
Sorry, but as I said previously, I'm no Java developper (I'm not even using eclipse for this patch). Do you mean something like
`logBasic ( BaseMessages.getString ( PKG ("Hit rate limiting. Sleeping ") + retryAfter + "s")`

Reply:
No worries, it would look something like the line below, and can be reused in most of the instances I've commented on.

```
logBasic( BaseMessages.getString( PKG, "ZendeskInput.Info.RateLimited", retryAfter );
```

You'd also need to add an entry into https://github.com/matthewtckr/pdi-zendesk-plugin/blob/master/src/main/resources/org/pentaho/di/trans/steps/zendesk/messages/messages_en_US.properties, something like:

```
ZendeskInput.Info.RateLimited=Zendesk API responded with rate limit.  Waiting {0} seconds before continuing.
```

This allows other developers to translate the English messages once into another language, and the translations would apply uniformly in all steps.

If you're not comfortable with the request, I can work on the remaining items later this week.